modules/metastorage-server/src/main/java/org/apache/ignite/internal/metastorage/server/RocksDBKeyValueStorage.java

```diff
@@ -0,0 +1,1249 @@
+            this.p = p;
+            this.lastRetRev = rev - 1;
+            this.it = createIterator();
+        }
+
+        /** {@inheritDoc} */
+        @Override public boolean hasNext() {
+            return it.hasNext();
+        }
+
+        /** {@inheritDoc} */
+        @Override public WatchEvent next() {
+            return it.next();
+        }
+
+        /** {@inheritDoc} */
+        @Override public void close() throws Exception {
+            IgniteUtils.closeAll(Set.of(nativeIterator, options));
+        }
+
+        /** {@inheritDoc} */
+        @NotNull
+        @Override public Iterator<WatchEvent> iterator() {
+            return it;
+        }
+
+        @NotNull
+        Iterator<WatchEvent> createIterator() {
+            return new Iterator<>() {
+                /** {@inheritDoc} */
+                @Override public boolean hasNext() {
+                    Lock lock = rwLock.readLock();
+                    lock.lock();
+                    try {
+                        if (nextRetRev != -1)
+                            return true;
+
+                        while (true) {
+                            long curRev = lastRetRev + 1;
+
+                            byte[] revisionPrefix = new byte[Long.BYTES];
+                            ByteUtils.toBytes(curRev, revisionPrefix, 0, Long.BYTES);
+
+                            boolean empty = true;
+
+                            if (nativeIterator == null)
+                                nativeIterator = db.newIterator(options);
+
+                            for (nativeIterator.seek(revisionPrefix); nativeIterator.isValid(); nativeIterator.next()) {
+                                empty = false;
+
+                                byte[] key = rocksKeyToBytes(nativeIterator.key());
+
+                                if (p.test(key)) {
+                                    nextRetRev = curRev;
+
+                                    return true;
+                                }
+                            }
+
+                            if (empty) {
+                                nativeIterator.close();
+
+                                nativeIterator = null;
+
+                                return false;
+                            }
+
+                            lastRetRev++;
+                        }
+                    }
+                    finally {
+                        lock.unlock();
+                    }
+                }
+
+                /** {@inheritDoc} */
+                @Override public WatchEvent next() {
+                    Lock lock = rwLock.readLock();
+                    lock.lock();
+                    try {
+                        while (true) {
+                            if (nextRetRev != -1) {
+                                boolean empty = true;
+
+                                List<EntryEvent> evts = new ArrayList<>();
+
+                                for (; nativeIterator.isValid(); nativeIterator.next()) {
```

Comment:
`nativeIterator` can be `null`, should we add an assertion?

Reply:
I don't see how it could be null

Reply:
Idea says that it can =)

Reply:
`nativeIterator` not-nullability is guarded by `nextRetRev != -1`, but `nextRetRev` is not -1 if `p.test(key)` is true, so I think that's why IDEA can't infer that `nativeIterator` can't be null at this moment.

Reply:
maybe we can add an assertion just to shut it up?